content/post/ddos-detection.md

```diff
@@ -0,0 +1,419 @@
+    And likewise, in order to change from "under attack" to "healthy", there must be at least 10 requests in the current prediction's sample.
+6. Whenever a server's status changes from either "healthy" to "under attack" and vice versa, we notify a consumer via a message over TCP, noting the server's address, the new status, and the time at which it changed.
+
+## The Application in Wallaroo
+
+![application diagram][application diagram]
+
+Our application's input will be a stream of request events, containing a timestamp, a client IP address, a server IP address, and a resource identifier, recorded as a JSON document.
+
+```
+...
+{"timestamp":1509494400.0,"client":"191.115.118.144","resource":"f06b667efd","server":"23.133.98.170"}
+{"timestamp":1509494400.001,"client":"212.177.244.108","resource":"8c4af94440","server":"187.224.241.67"}
+{"timestamp":1509494400.002,"client":"229.28.98.137","resource":" 594e9919a","server":"224.217.101.60"}
+...
+```
+
+In this example, we will use synthetic data generated with a python script that is included along with the demo code at [ddos-detection].
+The data has the following characteristics:
+
+1. 100 servers, serving 10,000 resources.
+2. 10 seconds of normal traffic, with 1,000 clients making 1,000 requests/second, distributed uniformly over all the servers.
+3. 10 seconds of attack traffic, with 100,000 clients making 25,000 requests/second, with 90% of the traffic hitting only 10 servers.
+4. 10 seconds of normal traffic, with 1,000 clients making 1,000 requests (in total) per second, distributed uniformly over all the servers.
+
+To produce a similar set, run
+
+```bash
+python data_gen.py --clients 1000 --attack-clients 100000 \
+  --requests 1000 --attack-requests 25000 --loaded-weight 0.9 \
+  --file data.json
+```
+
+To consume this stream, we will need to decode the JSON documents with a Wallaroo `Decoder`, using the standard library's `json` module:
+
+```python
+class Decoder(object):
+    def header_length(self):
+        return 4
+
+    def payload_length(self, bs):
+        return struct.unpack(">I", bs)[0]
+
+    def decode(self, bs):
+        return json.loads(bs)
+```
+
+As each server's status is computed independently of the others, we need a way to choose the partition for each incoming record. We could create a partition for each unique IP address, but that would require a partition index list the size of _2^32 - 1_ (assuming only IPv4 addresses), which will be a massive waste of memory given that we only have a small number of servers to monitor (certainly much less than 4.29 billion!). Instead, we will divide the IP range into _K_ even segments to be our partitions. We can then convert the string IP address in each record from the form of `A.B.C.D` to an integer, and determine the index of the segment it is in, which will become its partition identifier:
```

Comment:
There's an assumption in here that I realized that 2^32 - 1 is 4.29 billion. Probably not obvious to everyone. Making it explicit would be good.

Reply:
Or maybe just removing the 4.29 billion reference altogether?